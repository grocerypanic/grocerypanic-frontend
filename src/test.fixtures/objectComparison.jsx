export const propCount = (hash, length) => {
  expect(Object.keys(hash).length).toBe(length);
};
