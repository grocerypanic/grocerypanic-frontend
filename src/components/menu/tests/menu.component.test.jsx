import React from "react";
import {
  render,
  cleanup,
  fireEvent,
  waitFor,
  act,
} from "@testing-library/react";
import { propCount } from "../../../test.fixtures/objectComparison";

import { Router } from "react-router-dom";
import { createBrowserHistory } from "history";

import Menu from "../menu.component";

import Header from "../../header/header.component";
import MenuItem from "../../menu-item/menu-item.component";
import Hint from "../../hint/hint.component";

import calculateMaxHeight from "../../../util/height";

jest.mock("../../header/header.component");
jest.mock("../../menu-item/menu-item.component");
jest.mock("../../hint/hint.component");
jest.mock("../../../util/height");

Header.mockImplementation(() => <div>MockComponent</div>);
MenuItem.mockImplementation(() => <div>MockComponent</div>);
Hint.mockImplementation(() => <div>MockComponent</div>);
calculateMaxHeight.mockImplementation(() => 200);

const props = {
  options: [{ name: "option1", location: "option1" }],
  headerTitle: "HeaderTitle",
  title: "Title",
  helpText: "HelpText",
};

describe("Setup environment", () => {
  let utils;
  let startPath = "/some/unmatched/path";
  const history = createBrowserHistory();
  beforeEach(() => {
    history.location.pathname = startPath;
    jest.clearAllMocks();
    utils = render(
      <Router history={history}>
        <Menu {...props} />
      </Router>
    );
  });

  afterEach(cleanup);

  it("should call the header component with the correct props", () => {
    expect(Header).toBeCalledTimes(1);
    const headerCall = Header.mock.calls[0][0];
    propCount(headerCall, 3);
    expect(headerCall.title).toBe(props.headerTitle);
    expect(headerCall.transaction).toBe(false);
    expect(headerCall.create).toBe(null);
  });

  it("should call the hint component with the correct props", () => {
    expect(Hint).toBeCalledTimes(1);
    const helpCall = Hint.mock.calls[0][0];
    propCount(helpCall, 1);
    expect(helpCall.children).toBe(props.helpText);
  });

  it("should call the menu item with the correct props", () => {
    expect(MenuItem).toBeCalledTimes(1);
    const menuItemCall = MenuItem.mock.calls[0][0];
    propCount(menuItemCall, 3);
    expect(menuItemCall.name).toBe(props.options[0].name);
    expect(menuItemCall.location).toBe(props.options[0].location);
    expect(menuItemCall.choose).toBeInstanceOf(Function);
  });

  it("should call calculateMaxHeight on render", () => {
    expect(calculateMaxHeight).toBeCalledTimes(1);
  });

  it("clicking on option1, should change the page accordingly", async (done) => {
    expect(MenuItem).toBeCalledTimes(1);
    const choose = MenuItem.mock.calls[0][0].choose;
    const newPath = "/a/new/path";

    await waitFor(() => expect(history.location.pathname).toBe(startPath));

    act(() => choose(newPath));

    await waitFor(() => expect(history.location.pathname).toBe(newPath));

    done();
  });

  it("a should call calculateMaxHeight again on a window resize", async (done) => {
    expect(calculateMaxHeight).toBeCalledTimes(1);
    fireEvent(window, new Event("resize"));
    await waitFor(() => expect(calculateMaxHeight).toBeCalledTimes(2));
    done();
  });

  it("should match the snapshot on file (styles)", () => {
    expect(utils.container).toMatchSnapshot();
  });
});
