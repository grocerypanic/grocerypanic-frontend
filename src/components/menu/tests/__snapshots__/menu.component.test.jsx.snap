// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Setup environment should match the snapshot on file (styles) 1`] = `
.c0 {
  margin-top: calc(50px + 26px);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .nav-tabs {
  height: 28px !important;
  font-size: 0.75em;
}

.c0 a[role="tab"] {
  padding-bottom: 0px;
  padding-top: 4px;
  padding-left: 8px;
  padding-right: 8px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 10px;
  border-style: solid;
  border-width: 1px;
  border-color: #000000;
}

.c2 {
  width: 100%;
  margin-left: 10px;
  margin-right: 10px;
  text-align: center;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  list-style-type: none;
  list-style-position: outside;
  overflow: scroll;
  max-height: 200px;
  border-style: solid;
  border-color: #808080;
  border-width: 10px;
}

.c4 {
  padding-left: 0px;
  min-width: 200px;
  background: #808080;
  padding: 0px;
}

.c4 li:last-child {
  margin-bottom: 0px;
}

<div>
  <div>
    MockComponent
  </div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 alert alert-success"
      >
        Title
      </div>
      <div
        class="c3 overflow-auto"
        size="200"
      >
        <div
          class="c4"
        >
          <div>
            MockComponent
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    MockComponent
  </div>
</div>
`;
